### PR TITLE
enhancement(vrl): Reverse and FlatMap

### PR DESCRIPTION
**Pull request by dbcfd**
Add vrl functions reverse and flat_map
 * reverse: Reverses an array
 * flat_map: Flatten an array of objects into inner arrays

### CODE REVIEW

**Comment by StephenWakely** [comment]
Thankyou for doing this!

I'm not sold on the need for the `flat_map` function.  Do you have a use case in mind? This function is the sort of thing I would like to implement once iteration support lands - see #9001. With iteration we could make a more general purpose function. We are hoping to implement that in the next quarter, so unless there is a pressing need I vote holding off until then.

If we do need it, it will need a different name as `flat_map` is something we should reserve for iteration.

I'll take a look at the code shortly. It is much appreciated!

File: lib/vrl/stdlib/src/reverse.rs
Comment (StephenWakely) [inline lib/vrl/stdlib/src/reverse.rs:49]: This could be improved, although admittedly through a slight rabbit hole of complexity.

Currently this is saying that the return is an array, but we don't know what the types of the array elements are.

There is one situation under which we do know what all the types are. If the parameter array is of type `Any -> x`, where x is some known type (`KindInfo`). For example, if the parameter is an array of integers - we know the result will also be an array of integers.

If you are willing to take on this challenge, you will need to add a function to `lib/vrl/compiler/src/type_def.rs`, perhaps also called `reverse`.  Have a look at the `restrict_array` function. `reverse` would be similar to this, but it would need to check that the restricted array only has a single element in its `BTreeMap`, and the `Index` would have to be `Index::Any`. If it does, we can return that array as the typedef.

If it doesn't, we can't guarantee what the types of the reversed array would be, so we should just return `Kind::all()` as the current implementation does.

The reason we can't guarantee the types for anything other than `Index::Any` is because the array can come from the  event passed into VRL. We don't know what the type of this array is, or how many elements it has. One could theoretically set a particular index in this array, which would result in the typedef for that array knowing the type for that given index. However, since we don't know the length of that array, when reversing we have no way of knowing where the value at that index will end up.

The return value could then be something like:

```rust
    fn type_def(&self, state: &state::Compiler) -> TypeDef {
        self.value.type_def(state).reverse()
    }
```

I hope that makes sense. The typedefs can get a bit hairy because we are trying to impose a type system on top of untyped values being passed in, so the type system has to cater for the unknowns as well as the knowns. Feel free to fire over any questions if you need!

File: lib/vrl/stdlib/src/reverse.rs
Comment (StephenWakely) [inline lib/vrl/stdlib/src/reverse.rs:69]: Theoretically this could have been:

```rust
            tdef: TypeDef::new().array_mapped::<(), Kind>(map! { (): Kind::Integer }),
```

However, the test here setups up the typedef for the array by defining each index as an integer:

```
1 -> Kind::Integer
2 -> Kind::Integer
3 -> Kind::Integer
```

So, the typedef here should be as it already is since theoretically VRL can't guarantee that at runtime there will not be a fourth element in the array that is a String.

**Review by StephenWakely — CHANGES_REQUESTED** [review]
Nice work! Many thanks.

We need to come to a consensus about `flat_map` as per my previous comment.

`reverse` is definitely something I would like to incorporate. I've left some notes about improving the type_def for `reverse` if you would be willing to have a look at it, that would be great!

**Comment by dbcfd** [comment]
Closing to split out to just reverse work.

**Pull request closed without merging**
